consensus/view_change_service.go

```diff
@@ -0,0 +1,524 @@
+package consensus
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"sync"
+	"time"
+
+	"github.com/harmony-one/harmony/crypto/bls"
+
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/harmony-one/harmony/core/types"
+
+	"github.com/ethereum/go-ethereum/common"
+	msg_pb "github.com/harmony-one/harmony/api/proto/message"
+	"github.com/harmony-one/harmony/consensus/quorum"
+	"github.com/harmony-one/harmony/consensus/signature"
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/p2p"
+)
+
+// MaxViewIDDiff limits the received view ID to only 100 further from the current view ID
+const MaxViewIDDiff = 100
+
+// TODO: view change message retry to accelerate the view change process
+
+// State contains current mode and current viewID
+type State struct {
+	mode    Mode
+	modeMux sync.RWMutex
+
+	// current view id in normal mode
+	// it changes per successful consensus
+	blockViewID uint64
+	cViewMux    sync.RWMutex
+
+	// view changing id is used during view change mode
+	// it is the next view id
+	viewChangingID uint64
+
+	viewMux sync.RWMutex
+}
+
+// Mode return the current node mode
+func (pm *State) Mode() Mode {
+	pm.modeMux.RLock()
+	defer pm.modeMux.RUnlock()
+	return pm.mode
+}
+
+// SetMode set the node mode as required
+func (pm *State) SetMode(s Mode) {
+	pm.modeMux.Lock()
+	defer pm.modeMux.Unlock()
+	pm.mode = s
+}
+
+// GetCurBlockViewID return the current view id
+func (pm *State) GetCurBlockViewID() uint64 {
+	pm.cViewMux.RLock()
+	defer pm.cViewMux.RUnlock()
+	return pm.blockViewID
+}
+
+// SetCurBlockViewID sets the current view id
+func (pm *State) SetCurBlockViewID(viewID uint64) {
+	pm.cViewMux.Lock()
+	defer pm.cViewMux.Unlock()
+	pm.blockViewID = viewID
+}
+
+// GetViewChangingID return the current view changing id
+// It is meaningful during view change mode
+func (pm *State) GetViewChangingID() uint64 {
+	pm.viewMux.RLock()
+	defer pm.viewMux.RUnlock()
+	return pm.viewChangingID
+}
+
+// SetViewChangingID set the current view changing id
+// It is meaningful during view change mode
+func (pm *State) SetViewChangingID(id uint64) {
+	pm.viewMux.Lock()
+	defer pm.viewMux.Unlock()
+	pm.viewChangingID = id
+}
+
+// GetViewChangeDuraion return the duration of the current view change
+// It increase in the power of difference betweeen view changing ID and current view ID
+func (pm *State) GetViewChangeDuraion() time.Duration {
+	pm.viewMux.RLock()
+	pm.cViewMux.RLock()
+	defer pm.viewMux.RUnlock()
+	defer pm.cViewMux.RUnlock()
+	diff := int64(pm.viewChangingID - pm.blockViewID)
+	return time.Duration(diff * diff * int64(viewChangeDuration))
+}
+
+// switchPhase will switch FBFTPhase to nextPhase if the desirePhase equals the nextPhase
+func (consensus *Consensus) switchPhase(subject string, desired FBFTPhase, override bool) {
```

Comment:
This should be in a general consensus file, not in view_change_service.

Reply:
will move it.